tutorial/lesson_16_rgb_generator.cpp

```diff
@@ -0,0 +1,190 @@
+
+        // We will compile this pipeline to handle memory layouts in
+        // several different ways, depending on the 'layout' generator
+        // param.
+        if (layout == Layout::Planar) {
+            // This pipeline as written will work with planar images only,
+            // meaning images in which each color channel is its own
+            // densely-packed array in memory. In terms of the strides
+            // described in lesson 10, the stride in x is one, the stride
+            // in y is the width of the image, and the stride in c is the
+            // width times the height.
+
+        } else if (layout == Layout::Interleaved) {
+            // Another common format is 'interleaved', in which the
+            // red, green, and blue values for each pixel occur next
+            // to each other in memory. In this case the stride in x
+            // is three, the stride in y is three times the width of
+            // the image, and the stride in c is one. We can tell
+            // Halide to assume (and assert) that this is the case
+            // for the input and output like so:
+
+            input
+                .set_stride(0, 3) // stride in dimension 0 (x) is three
+                .set_stride(2, 1); // stride in dimension 2 (c) is one
+
+            brighter.output_buffer()
+                .set_stride(0, 3)
+                .set_stride(2, 1);
+
+            // For interleaved layout, you may want to use a different
+            // schedule. We'll tell Halide to additionally assume and
+            // assert that there are three color channels, then
+            // exploit this fact to make the loop over 'c' innermost
+            // and unrolled.
+
+            input.set_bounds(2, 0, 3); // Dimension 2 (c) starts at 0 and has extent 3.
+            brighter.output_buffer().set_bounds(2, 0, 3);
+
+            // Move the loop over color channels innermost and unroll
+            // it.
+            brighter.reorder(c, x, y).unroll(c);
+
+        } else if (layout == Layout::Either) {
+            // We can also remove all constraints and compile a
+            // pipeline that will work with any memory layout. It will
+            // probably be slow, because all vector loads become
+            // gathers, and all vector stores become scatters.
+            input.set_stride(0, Expr()); // Use a default-constructed
+                                         // undefined Expr to mean
+                                         // there is no constraint.
+
+            brighter.output_buffer().set_stride(0, Expr());
+
+        } else if (layout == Layout::Specialized) {
+            // We can accept any memory layout with good performance
+            // by telling Halide to inspect the memory layout at
+            // runtime, and branch to different code depending on the
+            // strides it finds. First we relax the default constraint
+            // that stride(0) == 1:
```

Comment:
Worth pointing out more explicitly that this is a codesize-vs-flexibility tradeoff; you wouldn't want to use this approach if you knew the input would only ever be (say) planar